include/h2o/tunnel.h

```diff
@@ -27,9 +27,50 @@ extern "C" {
 #endif
 
 typedef struct st_h2o_tunnel_t h2o_tunnel_t;
+typedef struct st_h2o_tunnel_end_t h2o_tunnel_end_t;
+typedef void (*h2o_tunnel_end_open_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_end_t *end);
+typedef void (*h2o_tunnel_end_send_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_end_t *end, h2o_iovec_t *bufs, size_t bufcnt,
+                                        int is_final);
+typedef void (*h2o_tunnel_end_on_peer_send_complete_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_end_t *end, h2o_tunnel_end_t *peer);
+typedef void (*h2o_tunnel_end_close_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_end_t *end, const char *err);
 
-h2o_tunnel_t *h2o_tunnel_establish(h2o_context_t *ctx, h2o_socket_t *sock1, h2o_socket_t *sock2, h2o_timeout_t *timeout);
-void h2o_tunnel_break(h2o_tunnel_t *tunnel);
+struct st_h2o_tunnel_end_t {
```

Comment:
I would appreciate it if you could write down the lifecycle of the "end", possible as a doc-comment here, because that will give us a good understanding on how it should be used.